File: api-hub/technical/coingecko.py
"""
CoinGecko Python Utility - مكتبة pycoingecko المجانية
=====================================================
✅ مجاني بالكامل - بدون مفتاح API
⚠️ الحد: 10-30 طلب/دقيقة

استخدام:
    from coingecko import CryptoTracker
    
    tracker = CryptoTracker()
    price = tracker.get_price('bitcoin')
    trending = tracker.get_trending()
"""

from pycoingecko import CoinGeckoAPI
from typing import List, Dict, Any, Optional, Union
import time
from functools import lru_cache
import logging

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class CryptoTracker:
    """
    CryptoTracker - واجهة موحدة للحصول على بيانات العملات المشفرة
    يستخدم pycoingecko للتواصل مع CoinGecko API مجاناً
    """
    
    def __init__(self):
        """تهيئة CoinGecko API client"""
        self.cg = CoinGeckoAPI()
        self.last_request_time = 0
        self.min_request_interval = 2.5  # 2.5 ثانية بين الطلبات
        
    def _rate_limit(self):
        """Rate limiting لضمان عدم تجاوز حد API المجاني"""
        current_time = time.time()
        time_since_last = current_time - self.last_request_time
        
        if time_since_last < self.min_request_interval:
            sleep_time = self.min_request_interval - time_since_last
            logger.debug(f"Rate limiting: sleeping for {sleep_time:.2f}s")
            time.sleep(sleep_time)
        
        self.last_request_time = time.time()
    
    def ping(self) -> Dict[str, str]:
        """
        اختبار الاتصال بـ CoinGecko API
        
        Returns:
            dict: {'gecko_says': '(V3) To the Moon!'}
        """
        self._rate_limit()
        try:
            return self.cg.ping()
        except Exception as e:
            logger.error(f"Ping failed: {e}")
            return {"error": str(e)}
    
    def get_price(
        self, 
        coin_ids: Union[str, List[str]], 
        vs_currencies: Union[str, List[str]] = 'usd',
        include_market_cap: bool = False,
        include_24hr_vol: bool = False,
        include_24hr_change: bool = False,
        include_last_updated_at: bool = False
    ) -> Dict[str, Dict[str, float]]:
        """
        جلب سعر عملة أو عدة عملات
        
        Args:
            coin_ids: معرف العملة أو قائمة معرفات (مثل 'bitcoin' أو ['bitcoin', 'ethereum'])
            vs_currencies: العملة المقابلة (افتراضي 'usd')
            include_market_cap: إضافة القيمة السوقية
            include_24hr_vol: إضافة حجم التداول 24 ساعة
            include_24hr_change: إضافة تغيير السعر 24 ساعة
            include_last_updated_at: إضافة وقت آخر تحديث
            
        Returns:
            dict: {'bitcoin': {'usd': 45000, 'usd_market_cap': 850000000000, ...}}
            
        Example:
            >>> tracker.get_price('bitcoin', 'usd', include_24hr_change=True)
            {'bitcoin': {'usd': 45000, 'usd_24h_change': 2.5}}
        """
        self._rate_limit()
        try:
            return self.cg.get_price(
                ids=coin_ids,
                vs_currencies=vs_currencies,
                include_market_cap=include_market_cap,
                include_24hr_vol=include_24hr_vol,
                include_24hr_change=include_24hr_change,
                include_last_updated_at=include_last_updated_at
            )
        except Exception as e:
            logger.error(f"Failed to get price: {e}")
            return {"error": str(e)}
    
    @lru_cache(maxsize=1)
    def get_coins_list(self) -> List[Dict[str, str]]:
        """
        جلب قائمة كل العملات المدعومة
        (مع كاش لتقليل الطلبات - البيانات نادراً ما تتغير)
        
        Returns:
            list: [{'id': 'bitcoin', 'symbol': 'btc', 'name': 'Bitcoin'}, ...]
        """
        self._rate_limit()
        try:
            return self.cg.get_coins_list()
        except Exception as e:
            logger.error(f"Failed to get coins list: {e}")
            return []
    
    def get_coin_by_id(
        self, 
        coin_id: str,
        localization: bool = False,
        tickers: bool = False,
        market_data: bool = True,
        community_data: bool = False,
        developer_data: bool = False
    ) -> Dict[str, Any]:
        """
        جلب معلومات شاملة عن عملة محددة
        
        Args:
            coin_id: معرف العملة (مثل 'bitcoin')
            localization: تضمين الترجمات
            tickers: تضمين بيانات المنصات
            market_data: تضمين بيانات السوق (أسعار، قيمة سوقية، إلخ)
            community_data: تضمين بيانات المجتمع
            developer_data: تضمين بيانات المطورين
            
        Returns:
            dict: معلومات شاملة عن العملة
        """
        self._rate_limit()
        try:
            return self.cg.get_coin_by_id(
                id=coin_id,
                localization=localization,
                tickers=tickers,
                market_data=market_data,
                community_data=community_data,
                developer_data=developer_data
            )
        except Exception as e:
            logger.error(f"Failed to get coin by id: {e}")
            return {"error": str(e)}
    
    def get_coin_market_chart(
        self,
        coin_id: str,
        vs_currency: str = 'usd',
        days: Union[int, str] = 30
    ) -> Dict[str, List]:
        """
        جلب بيانات السوق التاريخية (للرسوم البيانية)
        
        Args:
            coin_id: معرف العملة
            vs_currency: العملة المقابلة
            days: عدد الأيام (1, 7, 14, 30, 90, 180, 365, 'max')
            
        Returns:
            dict: {
                'prices': [[timestamp, price], ...],
                'market_caps': [[timestamp, market_cap], ...],
                'total_volumes': [[timestamp, volume], ...]
            }
        """
        self._rate_limit()
        try:
            return self.cg.get_coin_market_chart_by_id(
                id=coin_id,
                vs_currency=vs_currency,
                days=days
            )
        except Exception as e:
            logger.error(f"Failed to get market chart: {e}")
            return {"error": str(e)}
    
    def get_coin_ohlc(
        self,
        coin_id: str,
        vs_currency: str = 'usd',
        days: int = 30
    ) -> List[List[float]]:
        """
        جلب بيانات OHLC (Open, High, Low, Close) للرسم البياني
        
        Args:
            coin_id: معرف العملة
            vs_currency: العملة المقابلة
            days: عدد الأيام (1, 7, 14, 30, 90, 180, 365)
            
        Returns:
            list: [[timestamp, open, high, low, close], ...]
        """
        self._rate_limit()
        try:
            return self.cg.get_coin_ohlc_by_id(
                id=coin_id,
                vs_currency=vs_currency,
                days=days
            )
        except Exception as e:
            logger.error(f"Failed to get OHLC: {e}")
            return []
    
    def get_top_coins(
        self,
        vs_currency: str = 'usd',
        limit: int = 100,
        page: int = 1,
        sparkline: bool = False,
        price_change_percentage: str = '24h'
    ) -> List[Dict[str, Any]]:
        """
        جلب أفضل العملات حسب القيمة السوقية
        
        Args:
            vs_currency: العملة المقابلة
            limit: عدد العملات (1-250)
            page: رقم الصفحة
            sparkline: تضمين sparkline data
            price_change_percentage: فترات التغيير ('1h', '24h', '7d', '14d', '30d', '200d', '1y')
            
        Returns:
            list: قائمة العملات مع بياناتها
        """
        self._rate_limit()
        try:
            return self.cg.get_coins_markets(
                vs_currency=vs_currency,
                order='market_cap_desc',
                per_page=limit,
                page=page,
                sparkline=sparkline,
                price_change_percentage=price_change_percentage
            )
        except Exception as e:
            logger.error(f"Failed to get top coins: {e}")
            return []
    
    def get_trending(self) -> Dict[str, List]:
        """
        جلب العملات الرائجة (Trending)
        
        Returns:
            dict: {
                'coins': [قائمة العملات الرائجة],
                'nfts': [قائمة NFTs الرائجة]
            }
        """
        self._rate_limit()
        try:
            return self.cg.get_search_trending()
        except Exception as e:
            logger.error(f"Failed to get trending: {e}")
            return {"coins": [], "nfts": []}
    
    def get_global_data(self) -> Dict[str, Any]:
        """
        جلب بيانات السوق العالمية
        
        Returns:
            dict: {
                'data': {
                    'active_cryptocurrencies': int,
                    'markets': int,
                    'total_market_cap': dict,
                    'total_volume': dict,
                    'market_cap_percentage': dict,
                    'market_cap_change_percentage_24h_usd': float,
                    ...
                }
            }
        """
        self._rate_limit()
        try:
            return self.cg.get_global()
        except Exception as e:
            logger.error(f"Failed to get global data: {e}")
            return {"error": str(e)}
    
    def search_coins(self, query: str) -> Dict[str, List]:
        """
        البحث عن العملات
        
        Args:
            query: نص البحث (اسم العملة أو رمزها)
            
        Returns:
            dict: {
                'coins': [نتائج البحث],
                'exchanges': [],
                'icos': [],
                'categories': [],
                'nfts': []
            }
        """
        self._rate_limit()
        try:
            return self.cg.search(query=query)
        except Exception as e:
            logger.error(f"Failed to search: {e}")
            return {"coins": []}
    
    def get_categories(self) -> List[Dict[str, Any]]:
        """
        جلب فئات العملات (DeFi, NFT, Gaming, إلخ)
        
        Returns:
            list: [
                {
                    'id': str,
                    'name': str,
                    'market_cap': float,
                    'market_cap_change_24h': float,
                    'volume_24h': float,
                    'top_3_coins': [str, str, str]
                },
                ...
            ]
        """
        self._rate_limit()
        try:
            return self.cg.get_coins_categories()
        except Exception as e:
            logger.error(f"Failed to get categories: {e}")
            return []
    
    def get_exchanges(self, limit: int = 100, page: int = 1) -> List[Dict[str, Any]]:
        """
        جلب قائمة منصات التداول
        
        Args:
            limit: عدد المنصات
            page: رقم الصفحة
            
        Returns:
            list: قائمة المنصات مع بياناتها
        """
        self._rate_limit()
        try:
            return self.cg.get_exchanges_list(per_page=limit, page=page)
        except Exception as e:
            logger.error(f"Failed to get exchanges: {e}")
            return []


# ============================================================================
# دوال مساعدة سريعة (Utility Functions)
# ============================================================================

def quick_price(coin_id: str) -> Optional[float]:
    """
    جلب سعر سريع لعملة واحدة
    
    Args:
        coin_id: معرف العملة (مثل 'bitcoin')
        
    Returns:
        float: السعر بالدولار أو None إذا فشل
        
    Example:
        >>> quick_price('bitcoin')
        45000.0
    """
    tracker = CryptoTracker()
    result = tracker.get_price(coin_id)
    
    if coin_id in result and 'usd' in result[coin_id]:
        return result[coin_id]['usd']
    return None


def quick_trending() -> List[str]:
    """
    جلب قائمة سريعة بأسماء العملات الرائجة
    
    Returns:
        list: أسماء العملات الرائجة
        
    Example:
        >>> quick_trending()
        ['bitcoin', 'ethereum', 'cardano', ...]
    """
    tracker = CryptoTracker()
    result = tracker.get_trending()
    
    if 'coins' in result:
        return [coin['item']['id'] for coin in result['coins']]
    return []


def quick_top_10() -> List[Dict[str, Any]]:
    """
    جلب أفضل 10 عملات
    
    Returns:
        list: أفضل 10 عملات مع بياناتها
    """
    tracker = CryptoTracker()
    return tracker.get_top_coins(limit=10)


# ============================================================================
# مثال للاستخدام (Usage Example)
# ============================================================================

if __name__ == "__main__":
    # تهيئة CryptoTracker
    tracker = CryptoTracker()
    
    # 1. اختبار الاتصال
    print("=" * 50)
    print("🔌 Ping Test")
    print("=" * 50)
    ping_result = tracker.ping()
    print(ping_result)
    print()
    
    # 2. جلب سعر Bitcoin
    print("=" * 50)
    print("💰 Bitcoin Price")
    print("=" * 50)
    btc_price = tracker.get_price('bitcoin', include_24hr_change=True)
    if 'bitcoin' in btc_price:
        print(f"Price: ${btc_price['bitcoin']['usd']:,.2f}")
        if 'usd_24h_change' in btc_price['bitcoin']:
            print(f"24h Change: {btc_price['bitcoin']['usd_24h_change']:.2f}%")
    print()
    
    # 3. جلب أفضل 5 عملات
    print("=" * 50)
    print("🏆 Top 5 Cryptocurrencies")
    print("=" * 50)
    top_coins = tracker.get_top_coins(limit=5)
    for i, coin in enumerate(top_coins, 1):
        print(f"{i}. {coin['name']} ({coin['symbol'].upper()})")
        print(f"   Price: ${coin['current_price']:,.2f}")
        print(f"   24h Change: {coin['price_change_percentage_24h']:.2f}%")
        print()
    
    # 4. جلب العملات الرائجة
    print("=" * 50)
    print("🔥 Trending Coins")
    print("=" * 50)
    trending = tracker.get_trending()
    if 'coins' in trending:
        for i, coin_data in enumerate(trending['coins'][:5], 1):
            coin = coin_data['item']
            print(f"{i}. {coin['name']} ({coin['symbol']})")
    print()
    
    # 5. البحث عن عملة
    print("=" * 50)
    print("🔍 Search: Ethereum")
    print("=" * 50)
    search_results = tracker.search_coins('ethereum')
    if 'coins' in search_results:
        for coin in search_results['coins'][:3]:
            print(f"- {coin['name']} ({coin['symbol']})")
    print()
    
    # 6. دوال مساعدة سريعة
    print("=" * 50)
    print("⚡ Quick Functions")
    print("=" * 50)
    print(f"BTC Quick Price: ${quick_price('bitcoin'):,.2f}")
    print(f"Trending: {', '.join(quick_trending()[:5])}")
    print()
    
    print("✅ All tests completed!")

File: src/lib/services/coingecko.ts
/**
 * CoinGecko API Service
 * خدمة جلب البيانات من CoinGecko API
 * ✅ مجاني بالكامل - بدون مفتاح API
 * ⚠️ الحد: 10-30 طلب/دقيقة
 */

// أنواع البيانات
export interface CoinMarketData {
  id: string;
  symbol: string;
  name: string;
  image: string;
  current_price: number;
  market_cap: number;
  market_cap_rank: number;
  fully_diluted_valuation: number | null;
  total_volume: number;
  high_24h: number;
  low_24h: number;
  price_change_24h: number;
  price_change_percentage_24h: number;
  price_change_percentage_7d_in_currency?: number;
  market_cap_change_24h: number;
  market_cap_change_percentage_24h: number;
  circulating_supply: number;
  total_supply: number | null;
  max_supply: number | null;
  ath: number;
  ath_change_percentage: number;
  ath_date: string;
  atl: number;
  atl_change_percentage: number;
  atl_date: string;
  sparkline_in_7d?: {
    price: number[];
  };
  last_updated: string;
}

export interface TrendingCoin {
  item: {
    id: string;
    coin_id: number;
    name: string;
    symbol: string;
    market_cap_rank: number;
    thumb: string;
    small: string;
    large: string;
    slug: string;
    price_btc: number;
    score: number;
    data?: {
      price: number;
      price_change_percentage_24h: { [key: string]: number };
      sparkline: string;
    };
  };
}

export interface GlobalData {
  data: {
    active_cryptocurrencies: number;
    upcoming_icos: number;
    ongoing_icos: number;
    ended_icos: number;
    markets: number;
    total_market_cap: { [key: string]: number };
    total_volume: { [key: string]: number };
    market_cap_percentage: { [key: string]: number };
    market_cap_change_percentage_24h_usd: number;
    updated_at: number;
  };
}

// نوع بيانات OHLC: [timestamp, open, high, low, close]
export type OHLCData = [number, number, number, number, number]; // [timestamp, open, high, low, close]

// نوع مختصر لبيانات العملة
export type CoinData = CoinMarketData;

// استخدام API Route الداخلية لتجنب مشاكل CORS
const isClient = typeof window !== "undefined";
const API_PROXY = "/api/coingecko";
const COINGECKO_API = "https://api.coingecko.com/api/v3";

// Cache للبيانات
const cache = new Map<string, { data: unknown; timestamp: number }>();
const CACHE_DURATION = 60 * 1000; // دقيقة واحدة للكاش
const CACHE_DURATION_LONG = 300 * 1000; // 5 دقائق للبيانات الثابتة

// Rate limiting - CoinGecko Free API: 10-30 طلب/دقيقة
let lastRequestTime = 0;
const MIN_REQUEST_INTERVAL = 2500; // 2.5 ثانية بين الطلبات (أكثر أماناً)

// دالة مساعدة للطلبات
async function fetchWithCache<T>(
  endpoint: string,
  cacheKey: string,
  cacheDuration: number = CACHE_DURATION
): Promise<T> {
  // التحقق من الكاش
  const cached = cache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < cacheDuration) {
    return cached.data as T;
  }

  // Rate limiting
  const now = Date.now();
  const timeSinceLastRequest = now - lastRequestTime;
  if (timeSinceLastRequest < MIN_REQUEST_INTERVAL) {
    await new Promise(resolve => setTimeout(resolve, MIN_REQUEST_INTERVAL - timeSinceLastRequest));
  }
  lastRequestTime = Date.now();

  try {
    // استخدام API Proxy في المتصفح، والاتصال المباشر في الخادم
    const url = isClient 
      ? `${API_PROXY}?endpoint=${encodeURIComponent(endpoint)}`
      : `${COINGECKO_API}${endpoint}`;
    
    const response = await fetch(url, {
      headers: { accept: "application/json" },
    });
    
    if (!response.ok) {
      throw new Error(`CoinGecko API Error: ${response.status}`);
    }

    const data = await response.json();
    
    // حفظ في الكاش
    cache.set(cacheKey, { data, timestamp: Date.now() });
    
    return data as T;
  } catch (error) {
    console.error("CoinGecko fetch error:", error);
    throw error;
  }
}

/**
 * جلب بيانات السوق للعملات
 */
export async function getMarketData(
  page: number = 1,
  perPage: number = 100,
  sparkline: boolean = true
): Promise<CoinMarketData[]> {
  const endpoint = `/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=${perPage}&page=${page}&sparkline=${sparkline}&price_change_percentage=24h,7d`;
  return fetchWithCache<CoinMarketData[]>(endpoint, `markets_${page}_${perPage}`);
}

/**
 * جلب العملات الرائجة
 */
export async function getTrendingCoins(): Promise<TrendingCoin[]> {
  const data = await fetchWithCache<{ coins: TrendingCoin[] }>(
    "/search/trending",
    "trending"
  );
  return data.coins;
}

/**
 * جلب البيانات العالمية
 */
export async function getGlobalData(): Promise<GlobalData> {
  return fetchWithCache<GlobalData>("/global", "global");
}

/**
 * جلب بيانات عملة محددة
 */
export async function getCoinData(coinId: string): Promise<CoinMarketData | null> {
  try {
    const markets = await getMarketData(1, 250);
    return markets.find((coin) => coin.id === coinId) || null;
  } catch {
    return null;
  }
}

/**
 * جلب أسعار OHLC
 */
export async function getOHLC(
  coinId: string,
  days: number = 7
): Promise<OHLCData[]> {
  const endpoint = `/coins/${coinId}/ohlc?vs_currency=usd&days=${days}`;
  return fetchWithCache<OHLCData[]>(endpoint, `ohlc_${coinId}_${days}`);
}

/**
 * البحث عن عملات
 */
export async function searchCoins(query: string): Promise<{
  coins: Array<{
    id: string;
    name: string;
    symbol: string;
    market_cap_rank: number;
    thumb: string;
  }>;
}> {
  const endpoint = `/search?query=${encodeURIComponent(query)}`;
  return fetchWithCache(endpoint, `search_${query}`);
}

/**
 * حساب قوة الاتجاه بناءً على البيانات
 */
export function calculateTrendStrength(coin: CoinMarketData): {
  strength: number;
  signal: "strong-buy" | "buy" | "neutral" | "sell" | "strong-sell";
  momentum: number;
} {
  const change24h = coin.price_change_percentage_24h || 0;
  const change7d = coin.price_change_percentage_7d_in_currency || 0;
  const volumeToMcap = coin.total_volume / coin.market_cap;
  
  // حساب الزخم
  let momentum = 50;
  momentum += change24h * 2;
  momentum += change7d * 0.5;
  momentum += volumeToMcap > 0.1 ? 10 : volumeToMcap > 0.05 ? 5 : 0;
  momentum = Math.max(0, Math.min(100, momentum));

  // حساب قوة الاتجاه
  let strength = 0;
  if (change24h > 0 && change7d > 0) strength += 30;
  if (change24h > 5) strength += 20;
  if (change7d > 10) strength += 20;
  if (volumeToMcap > 0.1) strength += 15;
  if (coin.ath_change_percentage > -20) strength += 15;
  strength = Math.min(100, strength);

  // تحديد الإشارة
  let signal: "strong-buy" | "buy" | "neutral" | "sell" | "strong-sell";
  if (change24h > 10 && change7d > 15) signal = "strong-buy";
  else if (change24h > 3 && change7d > 5) signal = "buy";
  else if (change24h < -10 && change7d < -15) signal = "strong-sell";
  else if (change24h < -3 && change7d < -5) signal = "sell";
  else signal = "neutral";

  return { strength, signal, momentum };
}

/**
 * تصنيف العملات حسب الأداء
 */
export function categorizeCoins(coins: CoinMarketData[]): {
  hot: CoinMarketData[];
  bullish: CoinMarketData[];
  bearish: CoinMarketData[];
  highVolume: CoinMarketData[];
} {
  const hot = coins
    .filter((c) => {
      const volumeRatio = c.total_volume / c.market_cap;
      return volumeRatio > 0.15 && Math.abs(c.price_change_percentage_24h) > 5;
    })
    .slice(0, 10);

  const bullish = coins
    .filter((c) => c.price_change_percentage_24h > 3)
    .sort((a, b) => b.price_change_percentage_24h - a.price_change_percentage_24h)
    .slice(0, 20);

  const bearish = coins
    .filter((c) => c.price_change_percentage_24h < -3)
    .sort((a, b) => a.price_change_percentage_24h - b.price_change_percentage_24h)
    .slice(0, 20);

  const highVolume = coins
    .sort((a, b) => b.total_volume - a.total_volume)
    .slice(0, 10);

  return { hot, bullish, bearish, highVolume };
}

// ============================================================================
// 🆕 دوال إضافية مستوحاة من pycoingecko
// ============================================================================

/**
 * ping - اختبار الاتصال
 */
export async function ping(): Promise<{ gecko_says: string }> {
  return fetchWithCache("/ping", "ping", CACHE_DURATION_LONG);
}

/**
 * جلب سعر عملة أو عدة عملات
 * مثال: getPrice(['bitcoin', 'ethereum'], ['usd', 'eur'])
 */
export async function getPrice(
  ids: string | string[],
  vsCurrencies: string | string[] = 'usd',
  options?: {
    includeMarketCap?: boolean;
    include24hrVol?: boolean;
    include24hrChange?: boolean;
    includeLastUpdatedAt?: boolean;
  }
): Promise<Record<string, Record<string, number>>> {
  const idsStr = Array.isArray(ids) ? ids.join(',') : ids;
  const vsStr = Array.isArray(vsCurrencies) ? vsCurrencies.join(',') : vsCurrencies;
  
  let endpoint = `/simple/price?ids=${idsStr}&vs_currencies=${vsStr}`;
  if (options?.includeMarketCap) endpoint += '&include_market_cap=true';
  if (options?.include24hrVol) endpoint += '&include_24hr_vol=true';
  if (options?.include24hrChange) endpoint += '&include_24hr_change=true';
  if (options?.includeLastUpdatedAt) endpoint += '&include_last_updated_at=true';
  
  return fetchWithCache(endpoint, `price_${idsStr}_${vsStr}`);
}

/**
 * جلب قائمة كل العملات المدعومة
 */
export async function getCoinsList(): Promise<Array<{
  id: string;
  symbol: string;
  name: string;
}>> {
  return fetchWithCache("/coins/list", "coins_list", CACHE_DURATION_LONG);
}

/**
 * جلب معلومات شاملة عن عملة محددة
 */
export async function getCoinById(coinId: string): Promise<any> {
  const endpoint = `/coins/${coinId}?localization=false&tickers=false&community_data=false&developer_data=false`;
  return fetchWithCache(endpoint, `coin_${coinId}`);
}

/**
 * جلب بيانات السوق التاريخية (chart data)
 */
export async function getCoinMarketChart(
  coinId: string,
  vsCurrency: string = 'usd',
  days: number | 'max' = 30
): Promise<{
  prices: [number, number][];
  market_caps: [number, number][];
  total_volumes: [number, number][];
}> {
  const endpoint = `/coins/${coinId}/market_chart?vs_currency=${vsCurrency}&days=${days}`;
  return fetchWithCache(endpoint, `chart_${coinId}_${days}`);
}

/**
 * جلب بيانات OHLC للرسم البياني
 */
export async function getCoinOHLC(
  coinId: string,
  vsCurrency: string = 'usd',
  days: 1 | 7 | 14 | 30 | 90 | 180 | 365 | 'max' = 30
): Promise<OHLCData[]> {
  const endpoint = `/coins/${coinId}/ohlc?vs_currency=${vsCurrency}&days=${days}`;
  return fetchWithCache(endpoint, `ohlc_${coinId}_${vsCurrency}_${days}`);
}

/**
 * جلب أفضل العملات حسب Market Cap
 */
export async function getTopCoins(
  limit: number = 100,
  page: number = 1,
  sparkline: boolean = false
): Promise<CoinMarketData[]> {
  const endpoint = `/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=${limit}&page=${page}&sparkline=${sparkline}&price_change_percentage=24h,7d`;
  return fetchWithCache(endpoint, `top_${limit}_${page}`);
}

/**
 * جلب الفئات (DeFi, NFT, etc)
 */
export async function getCategories(): Promise<Array<{
  id: string;
  name: string;
  market_cap: number;
  market_cap_change_24h: number;
  volume_24h: number;
  top_3_coins: string[];
}>> {
  return fetchWithCache("/coins/categories", "categories", CACHE_DURATION_LONG);
}

/**
 * جلب بيانات المنصات (Exchanges)
 */
export async function getExchanges(limit: number = 100): Promise<Array<{
  id: string;
  name: string;
  year_established: number;
  country: string;
  image: string;
  trust_score: number;
  trust_score_rank: number;
  trade_volume_24h_btc: number;
}>> {
  const endpoint = `/exchanges?per_page=${limit}`;
  return fetchWithCache(endpoint, `exchanges_${limit}`, CACHE_DURATION_LONG);
}

/**
 * جلب العملات الرائجة (Trending)
 */
export async function getSearchTrending(): Promise<{
  coins: TrendingCoin[];
  nfts: Array<{ id: string; name: string; symbol: string; thumb: string }>;
}> {
  return fetchWithCache("/search/trending", "trending");
}

/**
 * جلب بيانات السوق العالمية
 */
export async function getGlobalMarketData(): Promise<{
  total_market_cap: number;
  total_volume: number;
  market_cap_percentage: Record<string, number>;
  market_cap_change_percentage_24h: number;
  active_cryptocurrencies: number;
  markets: number;
}> {
  const data = await fetchWithCache<GlobalData>("/global", "global_market");
  return {
    total_market_cap: data.data.total_market_cap.usd,
    total_volume: data.data.total_volume.usd,
    market_cap_percentage: data.data.market_cap_percentage,
    market_cap_change_percentage_24h: data.data.market_cap_change_percentage_24h_usd,
    active_cryptocurrencies: data.data.active_cryptocurrencies,
    markets: data.data.markets
  };
}

// ============================================================================
// 🔧 Utility Class - مثل pycoingecko
// ============================================================================

/**
 * CryptoTracker class - واجهة موحدة مثل pycoingecko
 */
export class CryptoTracker {
  /**
   * اختبار الاتصال
   */
  async ping() {
    return ping();
  }

  /**
   * جلب سعر عملة
   */
  async getPrice(coinId: string) {
    const data = await getPrice(coinId, 'usd', {
      include24hrChange: true,
      includeMarketCap: true,
      include24hrVol: true
    });
    return data[coinId];
  }

  /**
   * جلب أسعار عدة عملات
   */
  async getPrices(coinIds: string[]) {
    return getPrice(coinIds, 'usd', {
      include24hrChange: true,
      includeMarketCap: true
    });
  }

  /**
   * جلب أفضل العملات
   */
  async getTopCoins(limit: number = 10) {
    return getTopCoins(limit);
  }

  /**
   * جلب معلومات عملة
   */
  async getCoinInfo(coinId: string) {
    return getCoinById(coinId);
  }

  /**
   * البحث عن عملة
   */
  async searchCoin(query: string) {
    const results = await searchCoins(query);
    return results.coins;
  }

  /**
   * جلب العملات الرائجة
   */
  async getTrending() {
    const data = await getSearchTrending();
    return data.coins;
  }

  /**
   * جلب بيانات السوق العالمية
   */
  async getGlobalData() {
    return getGlobalMarketData();
  }

  /**
   * جلب بيانات الرسم البياني
   */
  async getChartData(coinId: string, days: number = 30) {
    return getCoinMarketChart(coinId, 'usd', days);
  }

  /**
   * جلب OHLC
   */
  async getOHLC(coinId: string, days: 1 | 7 | 14 | 30 | 90 | 180 | 365 = 30) {
    return getCoinOHLC(coinId, 'usd', days);
  }
}

// تصدير instance جاهز للاستخدام
export const cryptoTracker = new CryptoTracker();
